### ISSUE TITLE
Publishing slow on custom software template

### ISSUE THREAD

**Issue by mikkeschiren** [backstage/backstage#5682]
## Expected Behavior

Using a software template I am adding a Drupal code base, at the publish step it should go fast to publish it to Github/Gitlab

## Current Behavior

For adding a vanilla Drupal, with no vendors or node_modules, it takes about 20 minutes to finish the publish step.

I do not know where the issue is. Slow connection or slow progress inside Backstage, or cookiecutter itself.

**Comment by benjdlambert** [comment]
Hey can we see the actual template you're trying to publish? and the `template.yaml` definition? It's quite hard to tell what's going on here with limited information

**Comment by mikkeschiren** [comment]
This is what I am using:
https://gitlab.wklive.net/wk-public/template-drupal

**Comment by mikkeschiren** [comment]
Any ideas? Time for publishing goes between 20-45 minutes, and that seems really weird.

**Comment by benjdlambert** [comment]
So looking at the template @mikkeschiren looks like there a lot of files and a lot of data in the skeleton. The files that are templated are going to get pushed to version control, and I think it's around 60mb right now that's made up of thousands of files, I think that this could be down to the amount of log messages produced from the git publish step as it creates a log line for each item.

I wonder if you run the app with `LOG_LEVEL=error` instead maybe it could help if the logger for the `scaffolder` inherits. 

I'd also recommend keeping the templates pretty thin, and letting `composer` do the heavy lifting to get the dependencies rather than storing them in git?

**Comment by benjdlambert** [comment]
Looks like we need to take a look at the performance of big templates either way, and see what's causing this.

**Comment by benjdlambert** [comment]
@OrkoHunter for visibility here, wondering if you and @mtlewis wanna pick this up and have a prod around what's taking so long? Could be log writes, or it could be the `git.add`? not sure.

**Comment by mtlewis** [comment]
Hi @mikkeschiren! Looking at the template you linked above, it seems like it no longer has this issue. It seems like perhaps you've switched to letting composer handle adding dependencies after the fact and thereby reduced the number of files in the template, is that right? 

In any case, we do still have an issue with publishing being slow for templates with many files. I've done some digging on this, but I'll add it as a separate comment to keep things a bit more digestible.

**Comment by mtlewis** [comment]
**TL;DR it looks like we can make publish fast by switching from running add one-by-one for each file in the template to running `git.add({ dir, filename: '.' })`. This gets us to a reasonable publish time even for templates with thousands of files. Have opened https://github.com/isomorphic-git/isomorphic-git/issues/1402 to try to confirm this is a safe way of adding all the files in the repo.**

To investigate this issue, I created an artificially-large template (10,000 text files) and tried to create a component from it. The fetch step took around 3m to complete using fetch:cookiecutter (and 2m20s using fetch:template in compat mode :tada:), which is probably acceptable given the number of files. 

The publish step on the other hand took more time than I had available to wait 😁. I suspect that this is due to the [code in helpers.ts](https://github.com/backstage/backstage/blob/master/plugins/scaffolder-backend/src/scaffolder/actions/builtin/helpers.ts#L91) which stages files one-by-one. This is the [recommended approach in the isomorphic-git docs](https://isomorphic-git.org/docs/en/faq#how-to-add-all-untracked-files-with-gitadd), but I was interested to see whether triggering all the adds synchronously would get things done faster in the end:

```diff
-  for (const filepath of paths) {
-    await git.add({ dir, filepath });
-  }
+  await Promise.all(paths.map(filepath => git.add({ dir, filepath })));
```

This caused all 10,000 log lines for adding files to be printed to the logs immediately (which I think eliminates slow logging from consideration as the source of the slowness), but it still didn't make any additional forward progress after starting to stage the files before I ran out of patience. I'm also hesitant to pursue this angle without some kind of confirmation that isomorphic-git can safely handle receiving multiple parallel requests to add files.

The last thing I tried was adding all the files in the repo using:

```diff
-  const paths = await globby(['./**', './**/.*', '!.git'], {
-    cwd: dir,
-    gitignore: true,
-    dot: true,
-  });
- 
-  for (const filepath of paths) {
-    await git.add({ dir, filepath });
-  }
+  await git.add({ dir, filepath: '.' }); 
```

This seems to work - it took about 30 seconds to complete on my Backstage instance with the same 10,000 file skeleton, which seems totally acceptable given the number of files. The big questions though are whether this approach differs meaningfully in behavior from the globby-based solution today, and whether it's appropriate to pass `.` as the filename in when running `add` in `isomorphic-git`. It's certainly not documented, but looking at [the source for add in isomorphic-git](https://github.com/isomorphic-git/isomorphic-git/blob/main/src/api/add.js#L65), there's certainly affordances for adding directories. I can't immediately think of any edge-cases that would mean that `.` doesn't work if `some-sub-directory` does work. I've opened https://github.com/isomorphic-git/isomorphic-git/issues/1402 in their repo to get their thoughts on this.

@benjdlambert @OrkoHunter for vis.

**Comment by benjdlambert** [comment]
Nice one @mtlewis, great write up and nice find!
